notary-server/src/middleware.rs

```diff
@@ -0,0 +1,98 @@
+use async_trait::async_trait;
+use axum::http::{header, request::Parts};
+use axum_core::extract::{FromRef, FromRequestParts};
+use tracing::{error, trace};
+
+use crate::{
+    domain::{auth::AuthorizationWhitelistRecord, notary::NotaryGlobals},
+    NotaryServerError,
+};
+
+/// Auth middleware to prevent DOS
+pub struct AuthorizationMiddleware;
+
+#[async_trait]
+impl<S> FromRequestParts<S> for AuthorizationMiddleware
+where
+    NotaryGlobals: FromRef<S>,
+    S: Send + Sync,
+{
+    type Rejection = NotaryServerError;
+
+    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
+        let notary_globals = NotaryGlobals::from_ref(state);
+        let Some(whitelist) = notary_globals.authorization_whitelist else {
+            trace!("Skipping authorization as whitelist is not set.");
+            return Ok(Self);
+        };
+        let auth_header = parts
+            .headers
+            .get(header::AUTHORIZATION)
+            .and_then(|value| std::str::from_utf8(value.as_bytes()).ok());
+
+        match auth_header {
+            Some(auth_header) => {
+                if api_key_is_valid(auth_header, whitelist) {
+                    trace!("Request authorized.");
+                    Ok(Self)
+                } else {
+                    let err_msg = "Invalid API key.".to_string();
+                    error!(err_msg);
+                    Err(NotaryServerError::UnauthorizedProverRequest(err_msg))
+                }
+            }
+            None => {
+                let err_msg = "Missing API key.".to_string();
+                error!(err_msg);
+                Err(NotaryServerError::UnauthorizedProverRequest(err_msg))
+            }
+        }
+    }
+}
+
+/// Helper function to check if an API key is in whitelist
+fn api_key_is_valid(api_key: &str, whitelist: Vec<AuthorizationWhitelistRecord>) -> bool {
```

Comment:
Great spot, indeed it is owned, let me change it to ref

Reply:
It's certainly better to pass by reference here, but I don't think that the `Vec` is cloned. Isn't that just a pointer copy, because `Vec` is heap allocated?

Reply:
You are right. It is cloned in `NotaryGlobals::from_ref`

Reply:
Fixed - uses `Arc` to make sure it's thread safe since there ll be references from different threads serving different requests